test/Test/Effect.hs

```diff
@@ -109,3 +113,15 @@ shouldMatchList' a b = sendIO $ shouldMatchList a b
 
 expectFailure' :: Has (Lift IO) sig m => Result a -> m ()
 expectFailure' res = sendIO $ expectFailure res
+
+-- | Succeeds if the action fails and fails otherwise.
+expectFatal' :: (Has Diagnostics sig m, Has (Lift IO) sig m) => m () -> m ()
+expectFatal' f = do
+  res <- recover f
+  case res of
+    Just _ -> expectationFailure' "Expected failure"
+    Nothing -> pure ()
+
```

Comment:
[nit]: Three notes:
- `recover` preserves things like warnings, which we should probably throw away, we can use `Control.Effect.Diagnostics.errorBoundary` for that.
- `Diagnostics` is an effect around `Result`, so we can leverage `expectFailure'`, `errorBoundary` also helps with that, since it returns `Result a`.
- `m () -> m ()` is restrictive, we throw away the inner value of the action under test anyway.  Let's make it `m a -> m ()`.  this prevents us from having to `void` the diagnostics we pass to this function.

```suggestion
expectFatal' :: (Has Diagnostics sig m, Has (Lift IO) sig m) => m a -> m ()
expectFatal' f = errorBoundary f >>= expectFailure'
```

Reply:
so much more terse.  Nice.